### PR TITLE
fix(install): clean stale parser staging files

### LINKED ISSUE #716: install(parser): process termination leaves staged shared libraries permanently in the data directory
## Summary

Parser compilation stages its output inside the persistent parser data directory, but stale staged files are never recovered. A crash, `SIGKILL`, power loss, or termination after the compiler creates output but before normal cleanup leaves `.LANG.PID.COUNTER.EXT.tmp` permanently under `$XDG_DATA_HOME/kakehashi/parser`.

## Evidence on current `origin/main`

- `src/install/parser.rs:335-345` creates the staged path directly in the persistent parser directory.
- Cleanup at `src/install/parser.rs:357` and `:363` only runs when control returns through the rename-error or compile-error paths.
- There is no startup/install scan for the generated parser temp naming scheme.
- Query installs explicitly provide the missing lifecycle: `TempQueryDirGuard` handles ordinary errors, while `recover_interrupted_query_installs` recognizes generated temp names and deletes those owned by dead processes (`src/install/queries.rs:423-504`, `:658-689`).

The compile subprocess deadline does not cover termination of the parent process itself. If the parent dies, Rust destructors and the explicit cleanup branches do not run.

## Impact

Repeated interrupted installs accumulate platform shared libraries in the user's persistent data directory. These files can be large and are invisible to normal language listing/uninstall because their final extension is `.tmp`, not the platform DLL extension. There is no CLI path that removes them.

## Expected

Before/while installing, safely identify kakehashi-owned parser staging names, preserve files belonging to live installer processes, and remove files belonging to dead processes. Normal cleanup should remain best-effort. The ownership/name parsing must not delete arbitrary user files.

## Scope

This is separate from archive temporary extraction and #600/#601; it concerns compiled parser artifacts deliberately staged in the persistent install directory.

### PR DESCRIPTION
**Pull request by atusy**
Closes #716.

## Problem

A killed parser installer can leave its compiled `.LANG.PID.COUNTER.EXT.tmp` artifact permanently in the persistent parser directory. Normal error cleanup never runs after a crash or `SIGKILL`, and later installs did not recover those artifacts.

## Change

- recognize only the exact kakehashi-owned parser staging filename shape
- on Unix, remove artifacts whose owner PID is no longer alive
- preserve artifacts belonging to live concurrent installers
- conservatively leave staging files untouched on platforms without a portable process-liveness check
- run recovery before the ordinary already-installed early return

## Verification

- `cargo test --lib install::parser::tests` (48 passed)
- `cargo fmt --check`
- `cargo clippy --lib --tests -- -D warnings` reaches only the two pre-existing `clippy::single_range_in_vec_init` failures in `src/language/injection/discovery.rs:1576,1586`; this PR does not touch that file

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved recovery from interrupted parser installations by cleaning only verifiable stale staging artifacts.
  * Prevented active or unrelated installation outputs from being removed during cleanup.
  * Improved reliability of separate-process parser compilation by safely inheriting staging locks on Unix.
* **Tests**
  * Expanded Unix test coverage for staging-lock inheritance and cleanup correctness across live vs dead processes and edge cases.

### CODE REVIEW

File: src/install/parser.rs [hunk 1]
```diff
@@ -376,6 +378,65 @@ pub fn install_parser(
...
+}
+
+#[cfg(unix)]
+fn cleanup_interrupted_parser_installs(parser_dir: &Path) -> Result<(), ParserInstallError> {
+    for entry in fs::read_dir(parser_dir)? {
+        let entry = entry?;
+        let path = entry.path();
+        let Some(pid) = staged_parser_pid(&path) else {
+            continue;
+        };
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs]: Fixed in 550c8f869. Cleanup now rejects non-regular entries using symlink metadata and the focused `cleanup_` suite covers staging-shaped directories and symlinks.

File: src/install/parser.rs [hunk 2]
```diff
@@ -939,6 +1000,45 @@ mod tests {
...
+    #[cfg(unix)]
+    #[test]
+    fn cleanup_removes_staged_parser_from_dead_process() {
+        let temp = tempdir().expect("temp dir");
+        let parser_dir = temp.path().join("parser");
+        fs::create_dir_all(&parser_dir).expect("create parser dir");
+        let mut pid = std::process::id().saturating_add(100_000);
+        while process_is_running(pid) {
+            pid = pid.saturating_add(1);
+        }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs]: Fixed in 05d7289d0. Tests now spawn and reap a short-lived child, giving a deterministic valid-but-dead PID without an unbounded search.

File: src/install/parser.rs [hunk 3]
```diff
@@ -376,6 +372,104 @@ pub fn install_parser(
...
+    if parts.next().is_some()
+        || !super::queries::is_safe_language_name(language)
+        || counter.is_empty()
+        || !counter.bytes().all(|byte| byte.is_ascii_digit())
+        || extension != std::env::consts::DLL_EXTENSION
+    {
+        return None;
+    }
+    pid.parse().ok()
+}
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs:472]: Fixed in 05d7289d0. Staging names are recognized only for canonical PIDs in 1..=i32::MAX; zero and oversized PID regression cases are preserved.

File: src/install/parser.rs [hunk 4]
```diff
@@ -376,6 +372,104 @@ pub fn install_parser(
...
+
+#[cfg(unix)]
+fn cleanup_interrupted_parser_installs(parser_dir: &Path) -> Result<(), ParserInstallError> {
+    for entry in fs::read_dir(parser_dir)? {
+        let entry = entry?;
+        let path = entry.path();
+        let Some(pid) = staged_parser_pid(&path) else {
+            continue;
+        };
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs]: Fixed in 550c8f869. Cleanup conservatively skips non-regular directory entries; focused directory and symlink tests pass.

File: src/install/parser.rs [hunk 5]
```diff
@@ -376,6 +372,104 @@ pub fn install_parser(
...
+            match fs::rename(&path, &claimed) {
+                Ok(()) => {}
+                Err(error) if error.kind() == std::io::ErrorKind::NotFound => continue,
+                Err(error) => return Err(ParserInstallError::IoError(error)),
+            }
+            match fs::remove_file(claimed) {
+                Ok(()) => {}
+                Err(error) if error.kind() == std::io::ErrorKind::NotFound => {}
+                Err(error) => return Err(ParserInstallError::IoError(error)),
+            }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs]: Addressed by 7308d9a58 and retained at current HEAD: claims use `hard_link`, whose `AlreadyExists` result retries without replacing the destination, rather than Unix rename replacement.

File: src/install/parser.rs [hunk 6]
```diff
@@ -376,6 +372,149 @@ pub fn install_parser(
...
+    let extension = parts.next()?;
+    if parts.next().is_some()
+        || !super::queries::is_safe_language_name(language)
+        || !is_canonical_decimal(counter)
+        || extension != std::env::consts::DLL_EXTENSION
+    {
+        return None;
+    }
+    pid.parse().ok()
+}
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs:472]: Addressed by 9f19a8f73 and strengthened in 05d7289d0: canonical decimal form is required and valid staging PIDs are limited to 1..=i32::MAX.

File: src/install/parser.rs [hunk 7]
```diff
@@ -376,6 +372,149 @@ pub fn install_parser(
...
+        Err(_) => true,
+    }
+}
+
+#[cfg(unix)]
+fn cleanup_interrupted_parser_installs(parser_dir: &Path) -> Result<(), ParserInstallError> {
+    for entry in fs::read_dir(parser_dir)? {
+        let entry = entry?;
+        let path = entry.path();
+        if let Some(pid) = cleanup_claim_pid(&path) {
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs]: Fixed in 550c8f869. Both the pre-open and post-open checks require a regular file, and final-component symlinks are refused with O_NOFOLLOW.

File: src/install/parser.rs [hunk 8]
```diff
@@ -376,6 +374,174 @@ pub fn install_parser(
...
+    }
+    let opened = file.metadata()?;
+    let current = match fs::metadata(path) {
+        Ok(metadata) => metadata,
+        Err(error) if error.kind() == std::io::ErrorKind::NotFound => return Ok(None),
+        Err(error) => return Err(ParserInstallError::IoError(error)),
+    };
+    if opened.dev() != current.dev() || opened.ino() != current.ino() {
+        return Ok(None);
+    }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs:543]: Fixed in 550c8f869. The post-lock identity check now uses `symlink_metadata`, the opened handle must be a regular file, and O_NOFOLLOW prevents following a swapped final-component symlink.

File: src/install/parser.rs [hunk 9]
```diff
@@ -376,6 +374,174 @@ pub fn install_parser(
...
+        Err(nix::errno::Errno::ESRCH) => false,
+        Err(_) => true,
+    }
+}
+
+#[cfg(unix)]
+fn cleanup_interrupted_parser_installs(parser_dir: &Path) -> Result<(), ParserInstallError> {
+    for entry in fs::read_dir(parser_dir)? {
+        let entry = entry?;
+        let path = entry.path();
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs:702]: Fixed in 550c8f869. Missing or permission-denied directory reads are tolerated and per-entry errors are skipped with `flatten`, matching best-effort recovery semantics.

File: src/install/parser.rs [hunk 10]
```diff
@@ -376,6 +374,207 @@ pub fn install_parser(
...
+        .read(true)
+        // The pathname can be replaced after symlink_metadata. Never block if
+        // an external actor swaps in a FIFO before open.
+        .custom_flags(nix::libc::O_NONBLOCK)
+        .open(path)
+    {
+        Ok(file) => file,
+        Err(error) if error.kind() == std::io::ErrorKind::NotFound => return Ok(None),
+        Err(error) => return Err(ParserInstallError::IoError(error)),
+    };
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs:527]: Fixed in 550c8f869. The cleanup open uses O_NOFOLLOW and treats ELOOP as a conservative skip; a staging-shaped symlink regression test verifies the link and target remain untouched.

File: src/install/parser.rs [hunk 11]
```diff
@@ -376,6 +374,207 @@ pub fn install_parser(
...
+        Ok(()) => {}
+        Err(error) if error.kind() == std::io::ErrorKind::WouldBlock => return Ok(None),
+        Err(error) => return Err(ParserInstallError::IoError(error)),
+    }
+    let opened = file.metadata()?;
+    let current = match fs::metadata(path) {
+        Ok(metadata) => metadata,
+        Err(error) if error.kind() == std::io::ErrorKind::NotFound => return Ok(None),
+        Err(error) => return Err(ParserInstallError::IoError(error)),
+    };
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs:540]: Fixed in 550c8f869. The post-lock path identity check now uses symlink_metadata rather than following a replacement symlink.

File: src/install/parser.rs [hunk 12]
```diff
@@ -376,6 +374,223 @@ pub fn install_parser(
...
+        Err(nix::errno::Errno::ESRCH) => false,
+        Err(_) => true,
+    }
+}
+
+#[cfg(unix)]
+fn cleanup_interrupted_parser_installs(parser_dir: &Path) -> Result<(), ParserInstallError> {
+    for entry in fs::read_dir(parser_dir)? {
+        let entry = entry?;
+        let path = entry.path();
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs]: Fixed in 550c8f869. Directory entries are consumed with `flatten`, so an individual entry error cannot abort installation.

File: src/install/parser.rs [hunk 13]
```diff
@@ -939,6 +1154,127 @@ mod tests {
...
+    #[cfg(unix)]
+    #[test]
+    fn cleanup_removes_staged_parser_from_dead_process() {
+        let temp = tempdir().expect("temp dir");
+        let parser_dir = temp.path().join("parser");
+        fs::create_dir_all(&parser_dir).expect("create parser dir");
+        let mut pid = std::process::id().saturating_add(100_000);
+        while process_is_running(pid) {
+            pid = pid.saturating_add(1);
+        }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs]: Fixed in 05d7289d0. Dead-PID fixtures now use a spawned, reaped child PID, eliminating the linear probe and saturation hazard.

File: src/install/parser.rs [hunk 14]
```diff
@@ -939,6 +1154,127 @@ mod tests {
...
+    #[cfg(unix)]
+    #[test]
+    fn cleanup_preserves_staging_file_claimed_by_another_cleaner() {
+        let temp = tempdir().expect("temp dir");
+        let parser_dir = temp.path().join("parser");
+        fs::create_dir_all(&parser_dir).expect("create parser dir");
+        let mut pid = std::process::id().saturating_add(100_000);
+        while process_is_running(pid) {
+            pid = pid.saturating_add(1);
+        }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs]: Fixed in 05d7289d0. Dead-PID fixtures now use a spawned, reaped child PID, eliminating the linear probe and saturation hazard.

File: src/install/parser.rs [hunk 15]
```diff
@@ -939,6 +1154,127 @@ mod tests {
...
+    #[cfg(unix)]
+    #[test]
+    fn cleanup_skips_non_file_staging_path() {
+        let temp = tempdir().expect("temp dir");
+        let parser_dir = temp.path().join("parser");
+        fs::create_dir_all(&parser_dir).expect("create parser dir");
+        let mut pid = std::process::id().saturating_add(100_000);
+        while process_is_running(pid) {
+            pid = pid.saturating_add(1);
+        }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs]: Fixed in 05d7289d0. Dead-PID fixtures now use a spawned, reaped child PID, eliminating the linear probe and saturation hazard.

File: src/install/parser.rs [hunk 16]
```diff
@@ -376,6 +374,237 @@ pub fn install_parser(
...
+    let name = path.file_name()?.to_str()?;
+    let rest = name.strip_prefix(".parser-cleanup.")?;
+    let mut parts = rest.split('.');
+    let pid = parts.next()?;
+    let counter = parts.next()?;
+    if parts.next().is_some() || !is_canonical_decimal(pid) || !is_canonical_decimal(counter) {
+        return None;
+    }
+    pid.parse().ok()
+}
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs:495]: Fixed in 19a22b46e. Cleanup claim PIDs now use the same canonical 1..=i32::MAX range as staging PIDs; focused parser cleanup tests pass.

[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs:495]: Fixed in 19a22b46e. Cleanup claim PIDs now use the same canonical 1..=i32::MAX range as staging PIDs; focused parser cleanup tests pass.

File: src/install/parser.rs [hunk 17]
```diff
@@ -376,6 +409,310 @@ pub fn install_parser(
...
+fn cleanup_claim_pid(path: &Path) -> Option<u32> {
+    let name = path.file_name()?.to_str()?;
+    let rest = name.strip_prefix(".parser-cleanup.")?;
+    let mut parts = rest.split('.');
+    let pid = parts.next()?;
+    let counter = parts.next()?;
+    if parts.next().is_some() || !is_canonical_decimal(pid) || !is_canonical_decimal(counter) {
+        return None;
+    }
+    pid.parse().ok()
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs]: Fixed in 19a22b46e. Cleanup claim PIDs now use the same canonical 1..=i32::MAX range as staging PIDs; focused parser cleanup tests pass.

File: src/install/parser.rs [hunk 18]
```diff
@@ -286,6 +289,9 @@ pub fn install_parser(
         return Err(ParserInstallError::AlreadyExists(parser_file));
     }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs:326]: Fixed in d36407793. Recovery now runs before the AlreadyExists return, and the regression test asserts that a rejected install still removes dead-process staging artifacts.

File: src/install/parser.rs [hunk 19]
```diff
@@ -286,6 +322,9 @@ pub fn install_parser(
...
 
+    fs::create_dir_all(&parser_dir)?;
+    cleanup_interrupted_parser_installs(&parser_dir)?;
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs]: Fixed in d36407793. Recovery now runs before the AlreadyExists return, and the regression test asserts that a rejected install still removes dead-process staging artifacts.

File: src/install/parser.rs [hunk 20]
```diff
@@ -939,6 +1245,242 @@ mod tests {
...
+            no_cache: false,
+            compile: ParserCompile::InProcess,
+        };
+
+        let result = install_parser("lua", &options);
+
+        assert!(
+            matches!(result, Err(ParserInstallError::AlreadyExists(path)) if path == installed)
+        );
+        assert!(stale.exists(), "no cleanup runs for a rejected install");
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs]: Fixed in d36407793. Recovery now runs before the AlreadyExists return, and the regression test asserts that a rejected install still removes dead-process staging artifacts.

File: src/install/parser.rs [hunk 21]
```diff
@@ -939,6 +1245,242 @@ mod tests {
...
+                .file_type()
+                .is_symlink(),
+            "staging-shaped symlink is preserved"
+        );
+        assert_eq!(fs::read(&target).expect("read target"), b"user data");
+    }
+
+    #[cfg(unix)]
+    #[test]
+    fn existing_parser_returns_before_stale_cleanup() {
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs]: Fixed in d36407793. Recovery now runs before the AlreadyExists return, and the regression test asserts that a rejected install still removes dead-process staging artifacts.

File: src/install/parser.rs [hunk 22]
```diff
@@ -286,6 +322,9 @@ pub fn install_parser(
         return Err(ParserInstallError::AlreadyExists(parser_file));
     }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs:326]: Fixed in d36407793. Recovery now runs before the AlreadyExists return, and the regression test asserts that a rejected install still removes dead-process staging artifacts.

[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs:326]: Fixed in d36407793. Recovery now runs before the AlreadyExists return, and the regression test asserts that a rejected install still removes dead-process staging artifacts.

File: src/install/parser.rs [hunk 23]
```diff
@@ -376,6 +409,320 @@ pub fn install_parser(
...
+            .custom_flags(nix::libc::O_NOFOLLOW)
+            .open(&marker)
+            .and_then(|mut file| {
+                file.write_all(PARSER_CLEANUP_MARKER)?;
+                file.sync_all()
+            });
+        if let Err(error) = marker_result {
+            let _ = fs::remove_dir(&candidate);
+            return Err(ParserInstallError::IoError(error));
+        }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs:574]: Fixed in 82e340340. A failed marker write/sync now unlinks the possibly partial owner file before removing the claim directory.

File: src/install/parser.rs [hunk 24]
```diff
@@ -939,6 +1168,168 @@ mod tests {
     use super::*;
     use tempfile::tempdir;
 
+    #[cfg(unix)]
+    fn terminated_child_pid() -> u32 {
+        let mut child = std::process::Command::new("true")
+            .spawn()
+            .expect("spawn short-lived child");
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs]: Fixed in 62d4642cb (and silenced in 46f8bd521). The dead-PID fixture now spawns the current test executable with --help, avoiding an external PATH dependency.

File: src/install/parser.rs [hunk 25]
```diff
@@ -376,6 +374,237 @@ pub fn install_parser(
...
+        ));
+        match fs::hard_link(path, &candidate) {
+            Ok(()) => {
+                match fs::remove_file(path) {
+                    Ok(()) => {}
+                    Err(error) if error.kind() == std::io::ErrorKind::NotFound => {}
+                    Err(error) => return Err(ParserInstallError::IoError(error)),
+                }
+                return Ok(Some(candidate));
+            }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs:594]: Addressed by the current directory-claim implementation (42bad234d and later hardening): the stale pathname is atomically renamed into a private claim directory, and removal is best-effort through the directory descriptor. The prior hard-link/remove error path no longer exists.

File: src/install/parser.rs [hunk 26]
```diff
@@ -939,6 +1289,288 @@ mod tests {
...
+    #[cfg(unix)]
+    #[test]
+    fn staging_lock_is_inherited_through_exec() {
+        let temp = tempdir().expect("temp dir");
+        let path = temp.path().join("staging");
+        let lock = fs::File::create(&path).expect("create staging file");
+        lock.lock_exclusive().expect("lock staging file");
+        let mut cmd = Command::new("sh");
+        cmd.arg("-c").arg("sleep 30");
+        inherit_staging_lock(&mut cmd, &lock);
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs]: Fixed in 62c9b97ac by spawning `sleep` directly. The focused inherited-lock regression test and `cargo fmt --check` pass.

**Comment by atusy** [comment]
@copilot review

File: src/install/parser.rs [hunk 27]
```diff
@@ -376,6 +409,323 @@ pub fn install_parser(
...
+    let rest = name.strip_prefix(".parser-cleanup.")?;
+    let mut parts = rest.split('.');
+    let pid = parts.next()?;
+    let counter = parts.next()?;
+    if parts.next().is_some() || !is_canonical_decimal(pid) || !is_canonical_decimal(counter) {
+        return None;
+    }
+    let pid = pid.parse().ok()?;
+    (1..=i32::MAX as u32).contains(&pid).then_some(pid)
+}
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs:495]: Fixed in 4358ce5c0. Both staging and claim counters must parse as u64.

File: src/install/parser.rs [hunk 28]
```diff
@@ -939,6 +1289,288 @@ mod tests {
...
+            std::env::consts::DLL_EXTENSION
+        ));
+        fs::write(&staged, b"unreadable lookalike").expect("write staging file");
+        fs::set_permissions(&staged, fs::Permissions::from_mode(0o000))
+            .expect("remove read permission");
+
+        cleanup_interrupted_parser_installs(&parser_dir).expect("cleanup remains best-effort");
+
+        assert!(staged.exists(), "unreadable staging file is preserved");
+    }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs:1522]: Fixed in 4358ce5c0. The permission test now skips when the runner can still open the mode-000 file.

File: src/install/parser.rs [hunk 29]
```diff
@@ -376,6 +409,323 @@ pub fn install_parser(
...
+            Err(_) => {
+                let _ = fs::remove_file(&marker);
+                let _ = fs::remove_dir(&candidate);
+                return Ok(None);
+            }
+        }
+    }
+}
+
+const PARSER_CLEANUP_MARKER: &[u8] = b"kakehashi parser cleanup v1\n";
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs:610]: Fixed in 4358ce5c0. PARSER_CLEANUP_MARKER is now cfg(unix).

File: src/install/parser.rs [hunk 30]
```diff
@@ -376,6 +409,323 @@ pub fn install_parser(
...
+    }
+    Ok(())
+}
+
+#[cfg(not(unix))]
+fn cleanup_interrupted_parser_installs(_parser_dir: &Path) -> Result<(), ParserInstallError> {
+    // There is no portable process-liveness API. Keep possibly live staging
+    // files instead of risking deletion of another installer process's output.
+    Ok(())
+}
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs:733]: Handled by the dependency child PR #724, which adds the Windows cleanup and watchdog protocol. This parent intentionally establishes the Unix protocol first.

File: src/install/parser.rs [hunk 31]
```diff
@@ -376,6 +409,323 @@ pub fn install_parser(
...
+            if !process_is_running(pid) {
+                remove_stale_cleanup_claim(&path);
+            }
+            continue;
+        }
+        let Some(pid) = staged_parser_pid(&path) else {
+            continue;
+        };
+        if !process_is_running(pid) {
+            // Claim the stale pathname atomically before deleting it. A new
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs:713]: Declining the age heuristic: deleting while the embedded PID is live can destroy an active long compile. Conservative false retention under PID reuse is the required safety tradeoff and is eventually revisited.

File: src/install/parser.rs [hunk 32]
```diff
@@ -376,6 +409,323 @@ pub fn install_parser(
...
+            .and_then(|mut file| {
+                file.write_all(PARSER_CLEANUP_MARKER)?;
+                file.sync_all()
+            });
+        if let Err(error) = marker_result {
+            // `write_all` or `sync_all` may fail after creating a partial
+            // marker. Remove it before the now-empty claim directory.
+            let _ = fs::remove_file(&marker);
+            let _ = fs::remove_dir(&candidate);
+            return Err(ParserInstallError::IoError(error));
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/parser.rs:573]: Declining removal without an owner marker: an empty lookalike claim is not proven protocol-owned, and the existing regression requires preserving unowned claim directories. Deletion safety takes precedence over reclaiming this tiny crash window.

**Comment by atusy** [comment]
@copilot review
